extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -378,37 +430,50 @@ public boolean apply(Throwable input)
               }
             }
 
-            final long nextOffset = record.offset() + 1;
-            final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(record.partition());
-
-            nextOffsets.put(record.partition(), nextOffset);
+            nextOffsets.put(record.partition(), record.offset() + 1);
+          }
 
-            if (nextOffset == endOffset && assignment.remove(record.partition())) {
-              log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
-              assignPartitions(consumer, topic, assignment);
-              stillReading = !assignment.isEmpty();
-            }
+          if (nextOffsets.get(record.partition()).equals(endOffsets.get(record.partition()))
+              && assignment.remove(record.partition())) {
+            log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
+            assignPartitions(consumer, topic, assignment);
+            stillReading = ioConfig.isPauseAfterRead() || !assignment.isEmpty();
           }
         }
       }
 
-      // Persist pending data.
-      final Committer finalCommitter = committerSupplier.get();
-      driver.persist(finalCommitter);
+      // if we didn't actually generate any segments, don't try to publish as this will throw errors on the overlord
+      if (driver.getActiveSegmentCount() == 0) {
+        log.info("No segments were created by this task");
+        stopGracefully();
+      } else {
+        // Persist pending data.
+        driver.persist(committerSupplier.get());
+      }
 
-      publishing = true;
-      if (stopping) {
-        // Stopped gracefully. Exit code shouldn't matter, so fail to be on the safe side.
-        return TaskStatus.failure(getId());
+      if (stopRequested && !publishOnStop) {
+        throw new InterruptedException("stopping without publishing");
       }
 
+      status = Status.PUBLISHING;
       final TransactionalSegmentPublisher publisher = new TransactionalSegmentPublisher()
       {
         @Override
         public boolean publishSegments(Set<DataSegment> segments, Object commitMetadata) throws IOException
         {
-          // Sanity check, we should only be publishing things that match our desired end state.
-          if (!ioConfig.getEndPartitions().equals(((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS))) {
+          final KafkaPartitions finalPartitions = toolbox.getObjectMapper().convertValue(
+              ((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS),
+              KafkaPartitions.class
+          );
+
+          if (stopRequested) {
```

Comment:
Hmm, okay, if you're not using `stopGracefullyAndPublish` (and don't think you will need to use it) how about removing it? I think the unused code and potential skipping of sanity check are complexifying things, and if it isn't giving us value, we shouldn't do them.